avar2.md

```diff
@@ -27,24 +27,28 @@ Axis variation table:
 | `uint16`	| `<reserved>`	| Permanently reserved; set to zero. |
 | `uint16`	| `axisCount`	| The number of variation axes for this font. This must be the same number as axisCount in the 'fvar' table. |
 | `SegmentMaps`	| `axisSegmentMaps[axisCount]`	| The segment maps array — one segment map for each axis, in the order of axes specified in the 'fvar' table. |
-| `Offset32To<DeltaSetIndexMap>` | **`axisIdxMap`** | Offset from beginning of the table, to optional DeltaSetIndexMap storing variation index mapping. |
-| `Offset32To<ItemVariationStore>` | **`varStore`** | Offset from beginning of the table, to optional ItemVariationStore storing variations. |
+| `Offset32` | `axisIdxMapOffset` | Offset from beginning of the table to `axisIdxMap`. |
+| `Offset32` | `varStoreOffset` | Offset from beginning of the table to `varStore`. |
+| `DeltaSetIndexMap` | **`axisIdxMap`** | Indexing of axes used in `varStore`. |
+| `ItemVariationStore` | **`varStore`** | List of `variationRegions` and list of `ItemVariationData` structures including `deltaSets`. |
```

Comment:
I don't like this change. It implies a certain order of structs when none should be implied. Following how tables are specified in GSUB/GPOS/GDEF/COLR etc tables, we should NOT include the referred-to structs at all, just the offsets. And although foreign to the spec, I like the `Offset32To<>` format better. Otherwise you have to specify the type of the struct the offset points to, in the comments.

Reply:
Your point is well made about incorrectly implied ordering, something I addressed in my CFF2 spec rewrite with a note in the text (which is definitely not ideal). Yet the reversion removes some clarity about the new avar data. The previous `axisIdxMap` and `varStore` are not offsets.

It also seems potentially confusing in these cases (incl. GSUB/GPOS/GDEF/COLR) that the actual data is not listed as an item in the table. I note that some other tables use notation such as `uint8[dataLength]` to indicate arbitrary data, which also seems appropriate here. Offsets are not necessarily bound to the current table (see loca).

How about this?

---

|    |  |   |
| --- | --- | --- |
| `Offset32<DeltaSetIndexMap>` | `axisIdxMapOffset` | Offset from beginning of the table to `axisIdxMap`. |
| `Offset32To<ItemVariationStore>` | `varStoreOffset` | Offset from beginning of the table to `varStore`. |
| `uint8[n]` | binary data (*n* bytes) | Contains **`axisIdxMap`** and **`varStore`** in either order with no padding. |

Reply:
Maybe your naming struct offsets for their contents is reasonable... after all, in code we don’t append the string “pointer” to the name of every variable that is a pointer. Yet in a documentation of a binary spec, it seems important to be abundantly/redundantly clear about distinguishing data from offsets.

Reply:
Should we allow avar to be version 2 if it does not have axisIdxMap or varStore, thus zero’d offsets? Maybe not, so as to maximise compatibility with old avar processors.

Reply:
> The previous axisIdxMap and varStore are not offsets.

I don't understand what you mean.

> It also seems potentially confusing in these cases (incl. GSUB/GPOS/GDEF/COLR) that the actual data is not listed as an item in the table.

It would be impossible to list everything since  the number of objects is not fixed in general. Even here, you can't list all the ItemVariationStore's substructures.

> I note that some other tables use notation such as uint8[dataLength] to indicate arbitrary data, which also seems appropriate here. Offsets are not necessarily bound to the current table (see loca).

I think the fix here is to rename the struct to "avar version 2.0 header". That's all.

I don't like the Offset in the name, but am happy to go with it.

Please remove the uint8[n].

Reply:
A compiler can legitimately mix up axisIdxMap in between varStore's various structs. HarfBuzz repacker would do that.

Reply:
> Should we allow avar to be version 2 if it does not have axisIdxMap or varStore, thus zero’d offsets? Maybe not, so as to maximise compatibility with old avar processors.

Please no. I'm against such restrictions. Compilers should be smart. Forcing consumers of fonts to reject fonts isn't good user experience.

Reply:
> Forcing consumers of fonts to reject fonts isn't good user experience.

Why are consumers forced to reject such fonts? Do you mean font makers will be tempted to “upgrade” all their avar’d fonts to v2 with zero’d offsets, futzing up the fonts for users of v1-only systems?

Reply:
> A compiler can legitimately mix up axisIdxMap in between varStore's various structs. HarfBuzz repacker would do that.

Interesting point, I hadn't considered that the structs might be non-contiguous. I’ll remove wording about ordering of the two structs.

Reply:
>> Forcing consumers of fonts to reject fonts isn't good user experience.

> Why are consumers forced to reject such fonts?

I mean, I don't like requirements like "offset must not be zero", which essentially makes an otherwise understandable font illegal.

I am of the opinion that the spec should put minimal restrictions on what the compiler can do.